### DISCUSSION TITLE
New to Julia: edit a package module without using Revise or dev

### DISCUSSION THREAD

**Discussion by ghost** [CliMA/ClimaOcean.jl#274] [Announcements]
I want to edit a src file of a cloned ClimaOcean module from the rev="main" version without using Revise which requires using the dev version which is out of date.

Is there a link giving some straight forward instructions for reincluding the edited module so that it is in the standard place in the package (and not for example in Main) and not having to build a whole new environment of module dependencies for example.

How do I ensure the cloned version of the package is being used?

Thanks for your advice.

**Comment by glwagner** [comment]
You should be able to use `Pkg.develop` / `dev` for that, but you may also just do your work inside the `ClimaOcean.jl` repo, using the package environment. In both cases, you need to sure that you are on `main` within the `ClimaOcean` repo by writing `git checkout main; git pull`.

In other words, clone `ClimaOcean`, navigate into the repo, and then type `julia --project` to open Julia there. That will be using the `src` in the repo. I also suggest using Revise though because otherwise you will have to restart julia to run with any new code which will be slow.

This is the standard way to develop a Julia package, so its good to learn!

**Comment by glwagner** [comment]
I'll convert this to a discussion since it can't be resolved by any changes to the souce code and we can continue the conversation there!

**Comment by waywardpidgeon** [comment]
I have used these comments to edit some files from ClimaOcean, and written up a draft set of instructions which (after reviewing and fixing!) may make it easier for others to get started. Then again, I may have made this more complicated than it need be!!:

[PkgEditing.txt](https://github.com/user-attachments/files/17939937/PkgEditing.txt)

**Reply by glwagner** [reply]
PS it is discouraged to paste links to code --- better to just paste the code directly into the comment, like so:

```
Draft recipe for editing a package from git in the Windows power shell environment:

We use ExPkg as the example package ExPkg, could be ClimaOcean for example

Navigate to a convenient local directory in the power shell PS.

Download a copy of the package:
PS> git clone https://github.com/ExPkg.jl.git

Check to see you have the version you need to edit, in the case of ClimaOcean it should be main:
PS> git checkout main; git pull

Next move into the package directory in your own file system and stay there:
PS> cd EXPkg

Run Julia but don't use the flag --project
PS> julia

If needed, for example for ECCO files, establish your credentials:
julia> ENV["ECCO_USERNAME"]= "your_user_name"; ENV["ECCO_PASSWORD"] = "special_long complicated access_password" # this is set by Nasa... 
See ClimaOcean.jl/src/ECCO/README.md for instructions.

Turn on the development version:
In Windows 10 you can go
julia> using Pkg; Pkg.develop(".")
In Windows 11 however for some unknown reason you may  need to go into package mode and then
Pkg> dev .

Next switch into EXPkg:
julia> Pkg.activate("ExPkg")

Add any other packages used by ExPkg:
julia> Pkg.instantiate()

In addition you will  need to add the Revise package:
julia> Pkg.add("Revise")

Now load two packages explicitly:
julia> using Revise, EXPkg

You can now edit and save files such as file.jl  directly out of your cloned copy, e.g. edit src/file.jl and then
julia> include("src/file.jl")

Depending on the file structure, you may have to include some additional source files.
```

**Reply by glwagner** [reply]
These instructions look like the same instructions for linux or unix --- is there something specific here for powershell?

**Comment by waywardpidgeon** [comment]
Good. I have amended the instructions to be more generic. Are they ok?

**Reply by glwagner** [reply]
Can you paste your edits here?

**Comment by waywardpidgeon** [comment]
Here it is:

Draft recipe for editing code in a  package from git:
4 December 2024

We use ExPkg as the example package ExPkg, could be ClimaOcean for example. In the case of Windows,
the power shell PS is recommended.

Navigate to a convenient local directory in the power shell PS.

Download a copy of the package:
shell> git clone https://github.com/ExPkg.jl.git
In the case of ClimaOcean it is better to ensure the main branch is chosen, since that contains 
the most up-to-date code:

Check to see you have the version you need to edit:
shell> git checkout main; git pull

Next move into the package directory in your own file system and stay there:
shell> cd EXPkg

Run Julia but don't use the flag --project
shell> julia

If needed, for example for ECCO files, establish your credentials using .netrc in .Julia or within a Julia session going
julia> ENV["ECCO_USERNAME"]= "your_user_name"; ENV["ECCO_PASSWORD"] = "special_long complicated access_password"
This password is given to each user by Nasa. It only lasts a few months and needs to be reset.  See ClimaOcean.jl/src/ECCO/README.md for instructions also. 

Next turn on the development version:
In Windows 10 and Linux you can go
julia> using Pkg; Pkg.develop(".")
In Windows 11 however, for some unknown reason, you may  need to go into package mode and then
Pkg> dev .

Next switch into the environment for EXPkg:
julia> Pkg.activate("ExPkg")

Add any other packages used by ExPkg:
julia> Pkg.instantiate()

In addition you may  need to add the Revise package:
julia> Pkg.add("Revise")

Now load two packages explicitly:
julia> using Revise, EXPkg

You can now edit and save files, for example file.jl,  directly out of your cloned copy, e.g. edit src/file.jl and then
julia> include("src/file.jl")

Depending on the file structure, you may have to include some additional source files.

**Reply by glwagner** [reply]
A few comments:

* you should put `using Revise` in your `startup.jl` file, which is located in `.julia/config`. This is recomended by the Revise documentation.
* I never use `develop` and then `activate`. I just start with `julia --project` from the beginning.

Also, these instructions are appropriate for _developing_ ClimaOcean. If you only wish to interact with ClimaOcean as a user, the recommended workflow is different. For that you should create a new julia environment and add ClimaOcean to that environment, and develop new scripts. For that you can start with a script in `examples/` by copy-pasting it into the directory of your new environment.

**Comment by waywardpidgeon** [comment]
PS please amend the recipe by going 

shell> cd ExPkg
before 
shell> git checkout main; git pull
and NOT after!

**Reply by glwagner** [reply]
You can edit your comment. Do you know how to do that?

**Reply by ghost** [reply]
No!

**Reply by glwagner** [reply]
Here's a screenshot. Click the three dots to the top right (red arrow). That will open a drop down menu. Select "edit" from the menu (blue arrow).

![image](https://github.com/user-attachments/assets/8b403619-cde6-46cb-9ad0-47c37d5df8de)

It could help to read the github docs about commenting and markdown etc too!
